### PR TITLE
Add parameters to decorators

### PR DESCRIPTION
**Pull request by f0ffee**
New `command_wrapper` parameters:
- `required_rank`
- `allow_pm`
- `main_room_only`
- `parametrize_room`

New `route_wrapper` parameter:
- `required_rank`

Ke belo

### CODE REVIEW

**Comment by palt0** [comment]
I welcome these changes! In particular, I'll need `required_rank` for Parnassius#103.

I really dislike the implementation of `allow_pm`, though.
Adding different type hints to a class based on an attribute is arguably more confusing than just keeping the original explicit condition.
More importantly, the base logic becomes overcomplicated and *loses extensibility*: if we were to add an additional `Optional` parameter in the future, how would you proceed? Adding other `Message` type hints and `command_wrapper` overloads for all the new possible combinations is a bad idea.
Reverting back this change would also be annoying because we'd have to edit a lot of plugin modules again.

**TL;DR** I'd avoid implementing `allow_pm` if it requires introducing a new type hint. Yes, it works by miracle now, but in the long run the proposed structure is worse; the new syntax is probably harder to understand too. Feel free to disagree though! :D

**Comment by f0ffee** [comment]
Then how would you suggest we circumvent the linter errors?

**Comment by palt0** [comment]
Huh no idea, the above message is just a summary of the points I brought up in a private discussion, for future reference. I'll experiment a bit when I start working on #103 maybe, but I'm not really interested as of now, sorry!

As I wrote, _personally_ I'd avoid implementing `allow_pm` with the current solution and just stick to a regular `if` for that check (because I don't like having to use different and it worsens `plugins/__init__.py`). I think you both disagree and it's totally fine!
